### DISCUSSION TITLE
Issue with MIfitHMM or with my data

### DISCUSSION THREAD

**Discussion by hance-ellington** [bmcclintock/momentuHMM#83] [Q&A] [answered]
I am using momentuHMM package to work with coyote movement data that is irregularly sampled. I have had no problem using crawlWrap, prepData, and fitHMM to generate HMM using single imputation but I am having trouble implementing multiple imputation using MIfitHMM. Below is my code, and the two types of errors I have received.

**crwOut <- crawlWrap(data, timeStep = "2 hours", theta=c(6.855, -0.007), fixPar=c(NA,NA))
CBcoyData <- prepData(data=crwOut)**

**dist <- list(step = "gamma", angle = "wrpcauchy")
Par0_m2 <- list(step=c(50,700,50,350),angle=c(3,0.3,0.3,0.5))**

**m2 <- fitHMM(data = CBcoyData, nbStates = 2, dist = dist, Par0 = Par0_m2, estAngleMean = list(angle=TRUE))  ### single imputation**

**nfsFits2 <- MIfitHMM(crwOut, nSims = 2, nbStates = 2, dist = dist,  Par0 = Par0_m2, estAngleMean = list(angle = TRUE), retryFits = 30)) ### two imputations for testing**

_In MIfitHMM(crwOut, nSims = 2, nbStates = 2, dist = dist, Par0 = Par0_m2,  :
  prepData failed for imputation 1; Error in eval(.doSnowGlobals$expr, envir = .doSnowGlobals$exportenv): 
  crawl::crwPostIS error for individual 26.01; Error in sample.int(length(x), size, replace, prob): NA in probability vector
  Check crwPostIS arguments, crawl::crwMLE model fits, and/or consult crawl documentation._

  _prepData failed for imputation 2; Error in eval(.doSnowGlobals$expr, envir = .doSnowGlobals$exportenv): 
  crawl::crwPostIS error for individual 26.01; Error in sample.int(length(x), size, replace, prob): NA in probability vector
  Check crwPostIS arguments, crawl::crwMLE model fits, and/or consult crawl documentation._

My first thought was these errors might be linked to high missing rate or situations where a single fix is surrounded by a series of missing fixes. So I removed the individual listed above from data, then reran. I received the error above twice more, each time with a new offending individual (that I then removed). Then I got this error:

_Drawing 2 realizations from the position process using crawl::crwPostIS...
Error in { : task 20 failed - "missing value where TRUE/FALSE needed"_

I then looked at all individuals and removed 13 more individuals that had single fixes surrounded by series of missing fixes. Then I started getting this type error again:

_prepData failed for imputation X; Error in eval(.doSnowGlobals$expr, envir = .doSnowGlobals$exportenv): 
  crawl::crwPostIS error for individual XXXX; Error in sample.int(length(x), size, replace, prob): NA in probability vector_

I removed a few more individuals based on this error, but I now I am down to half my initial sample size and beginning to wonder if there is some other issue causing the errors.

Note I have also been successful with single imputation using 3 and 4 movement states but again was unsuccessful implementing multiple imputation (with the same error codes).  

Is there something wrong with my data or code?

**Comment by bmcclintock** [comment, marked as answer]
Although the error is occurring in MIfitHMM, this issue is more about model fitting and simulation in crawl. crawlWrap and the simulation step in MIfitHMM are really just wrappers for four functions in crawl (`crawl::crwMLE` and `crawl::crwPredict` in crawlWrap; `crawl::crwSimulator` and `crawl::crwPostIS` in MIfitHMM). If there is an issue with the output from any of these functions, then crawlWrap and/or MIfitHMM will fail. Thus it is important to get familiar with these crawl functions before attempting to use crawlWrap or MIfitHMM (with a crwData object).

In this particular case, it would seem that some of the individual model fits from crawlWrap are not reliable due to: 1) lack of convergence; and/or 2) parameters hitting boundaries (both of which can produce nonsensical point estimates and variances). Assuming the individual fits appear to have converged (this can be helped by using the 'attempts' and 'retryFits' arguments to better explore the parameter space), I suspect if you examine the crawlWrap output (`crwOut$crwFits`), the offending tracks will have estimates that are hitting boundaries. This does not affect single imputation because `crawl::crwPredict` does not rely on the variance estimates, but `crawl::crwSimulator` and `crawl::crwPostIS` need these variance estimates to draw realizations from the predicted path.

One common solution is to use the 'prior' argument to keep the parameters from wandering into the nether regions of the parameter space along boundaries.  Use of the 'prior' argument is described in the `crawl::crwMLE` documentation and [here](https://jmlondon.github.io/crawl-workshop/). For the log-scale beta parameter, London and Johnson currently suggest a normal prior with a mean of -4 and a standard deviation of 2. Based on your code, this could be specified in crawlWrap as:
```
ln.prior <- function(theta) dnorm(theta[2],-4,2,log=TRUE)
crwOut <- crawlWrap(data, 
                    timeStep = "2 hours", 
                    theta=c(6.855, -0.007), 
                    fixPar=c(NA,NA), 
                    prior=ln.prior)
```
So, in summary, in crawlWrap I'd suggest using: 1) 'attempts' and 'retryFits' to help with convergence; and 2) 'prior' to help with any boundary issues.

**Comment by hance-ellington** [comment]
Thank you, adding the prior argument improved the imputation process and allowed it to proceed with m=2 tests. I will update issue after testing with m=100.
